File: ignicao-api/src/clients/application/usecases/get-client.usecase.ts
import { ClientRepository } from 'src/clients/domain/repositories/client.repository';
import { UseCase } from 'src/shared/application/usecases/usecase';
import { ClientOutputEntity } from '../dto/client-output.dto';

export type Input = {
  id: string;
};

export class GetClientUsecase implements UseCase<Input, ClientOutputEntity> {
  constructor(private readonly clientRepository: ClientRepository) {}

  async execute(input: Input): Promise<ClientOutputEntity> {
    const entity = await this.clientRepository.findById(input.id);
    return entity.toJson();
  }
}

File: ignicao/lib/features/clients/data/usecases/get_client_usecase.dart
import 'package:dartz/dartz.dart';

import '../../../../shared/data/usecases/usecase.dart';
import '../../../../shared/errors/failures.dart';
import '../../domain/dto/get_client_dto.dart';
import '../../domain/entities/client_output.dart';
import '../../domain/repositories/client_repository.dart';

class GetClientUseCase implements UseCase<ClientOutputEntity?, GetClientDto> {
  GetClientUseCase(this.repository);

  final ClientRepository repository;

  @override
  Future<Either<Failure, ClientOutputEntity?>> call(GetClientDto params) async {
    return repository.getClient(params);
  }
}
